### ISSUE TITLE
Modify learning rate with multistep_optimizer

### ISSUE THREAD

**Issue by hoangcuong2011** [tensorflow/tensor2tensor#1769]
Greetings,

I found [multistep_optimizer](https://github.com/tensorflow/tensor2tensor/blob/2330203cb267fa4efc4525ad5c7dffe0a7d2f2fc/tensor2tensor/utils/multistep_optimizer.py) is an amazing class that helps me train large batch size. I, however, have an issue with this class in the sense that I also need to change the learning rate during training, and I found it is not easy to do so. 

In detail, I am trying to train the Transformer with the code I wrote from scratch. Training the model needs a good care of adapting learning rate based on the training steps. Normally with common optimizers like Adam, what I need to is just write a Callback and then modify lr as: `self.model.optimizer.lr = lrate  `. I however found this code does not work for this class (`AttributeError: 'TFOptimizer' object has no attribute 'lr'`)

So how to modify learning rate with this class?
Thx

**Comment by hoangcuong2011** [comment]
@martinpopel @lukaszkaiser I tagged your github account, hopefully to have your attention on my question. Many thanks!

**Comment by martinpopel** [comment]
I agree multistep optimizers (delayed updates) are very useful, but I have never used MultistepAdamOptimizer so far, so I am not sure with this answer.
MultistepAdamOptimizer is a subclass of tf.compat.v1.train.AdamOptimizer, so I guess it inherits `self._lr` and `self._lr_t`.
It is pity, T2T does not have multistep optimizer for sparse updates and for Adafactor.

**Comment by hoangcuong2011** [comment]
@martinpopel : Thx for your reply. When I change self.model.optimizer.lr to `self.model.optimizer._lr`or `self.model.optimizer._lr_t`, the error is still the same. So I guess it is not the way to fix the issue. 
BTW I am curious when sparse updates are used? I suppose if we call Adam, MultistepAdamOptimizer, or any common popular optimizers (SGD, RMSProb), it would call _apply_dense instead of _apply_sparse?

@all: After some effort I finished writing a class that is based on multistep_optimizer but the learning rate is changed inside the class according to Transformer paper. I put it here just in case anyone who needs to reuse. 

@lukaszkaiser and others who I noticed also contributed to multistep_optimizer class @lgeiger @rsepassi @fstahlberg @allenlavoie : I checked the code carefully and I believe my code is correct, but please help me by having some check if you can).

Thank you all
```

class MultistepAdamOptimizer(tf.compat.v1.train.AdamOptimizer):
    """Adam with SGD updates every n steps with accumulated gradients."""
    def __init__(self, learning_rate=0.001, beta1=0.9, beta2=0.999, epsilon=1e-8, use_locking=False, name="Adam", warmmup_steps =4000, d_model=512, n=1):
        super(MultistepAdamOptimizer, self).__init__(learning_rate=learning_rate, beta1=beta1, beta2=beta2, epsilon=epsilon, use_locking=use_locking, name=name)
        self._n = n  # Call Adam optimizer every n batches with accumulated grads
        self._n_t = None  # n as tensor
        self.constant_d_model_tensor = K.cast(d_model, 'float32')
        self.constant_minus05_tensor = K.cast(-0.5, 'float32')
        self.constant_minus15_tensor = K.cast(-1.5, 'float32')
        self.constant_warmup_tensor = K.cast(warmmup_steps, 'float32')

    def _create_slots(self, var_list):
        """Create slot variables for Adam with accumulated gradients."""
        # Create the beta1 and beta2 accumulators, and lambda_update on the same device as the first
        # variable. Sort the var_list to make sure this device is consistent across
        # workers (these need to go on the same PS, otherwise some updates are
        # silently ignored).
        first_var = min(var_list, key=lambda x: x.name)
        self._create_non_slot_variable(initial_value=self._beta1, name="beta1_power", colocate_with=first_var)
        self._create_non_slot_variable(initial_value=self._beta2, name="beta2_power", colocate_with=first_var)
        self._create_non_slot_variable(initial_value=0.00001, name="lambda_update", colocate_with=first_var)
        self._create_non_slot_variable(initial_value=0 if self._n == 1 else 1, name="iter", colocate_with=first_var)
        # Create slots for the first and second moments, as well as grad_acc.
        for v in var_list:
            self._zeros_slot(v, "m", self._name)
            self._zeros_slot(v, "v", self._name)
            self._zeros_slot(v, "grad_acc", self._name)
    def _get_iter_variable(self):
        graph = (None if tf.executing_eagerly() else tf.get_default_graph())
        return self._get_non_slot_variable("iter", graph=graph)

    def _prepare(self):
        super(MultistepAdamOptimizer, self)._prepare()
        self._n_t = tf.convert_to_tensor(self._n, name="n")

    def _apply_cond(self, apply_fn, grad, var, *args, **kwargs):
        """Apply conditionally if counter is zero."""
        grad_acc = self.get_slot(var, "grad_acc")

        def apply_adam(grad_acc, apply_fn, grad, var, *args, **kwargs):
            total_grad = (grad_acc + grad) / tf.cast(self._n_t, grad.dtype)
            adam_op = apply_fn(total_grad, var, *args, **kwargs)
            with tf.control_dependencies([adam_op]):
                grad_acc_to_zero_op = grad_acc.assign(tf.zeros_like(grad_acc), use_locking=self._use_locking)
            return tf.group(adam_op, grad_acc_to_zero_op)

        def accumulate_gradient(grad_acc, grad):
            assign_op = tf.assign_add(grad_acc, grad, use_locking=self._use_locking)
            return tf.group(assign_op)  # Strip return value
        return tf.cond(tf.equal(self._get_iter_variable(), 0),
            lambda: apply_adam(grad_acc, apply_fn, grad, var, *args, **kwargs),
            lambda: accumulate_gradient(grad_acc, grad))

    
    def _apply_sparse_shared(self, grad, var, indices, scatter_add):
        return self._apply_cond(super(MultistepAdamOptimizer, self)._apply_sparse_shared, grad, var,
        indices, scatter_add)

    def _apply_sparse(self, grad, var):
        # TODO(fstahlberg): Implement a sparse version
        tf.logging.warning("MultistepAdamOptimizer does not support sparse updates")
        dense_grad = tf.convert_to_tensor(grad)
        return self._apply_cond(self._apply_dense, dense_grad, var)

    def _resource_apply_sparse_duplicate_indices(self, grad, var, indices):
        tf.logging.warning("MultistepAdamOptimizer does not support sparse updates")
        # Note that conversion to a dense Tensor handles duplicate `indices`
        # correctly (summing them). A real sparse implementation will probably want
        # to override _resource_apply_sparse instead so it gets them de-duplicated
        # automatically.
        dense_grad = tf.convert_to_tensor(tf.IndexedSlices(values=grad, indices=indices, dense_shape=tf.shape(var)))
        return self._apply_cond(self._resource_apply_dense, dense_grad, var)        

    def _apply_dense(self, grad, var):
        m = self.get_slot(var, "m")
        v = self.get_slot(var, "v")
        beta1_power, beta2_power, lambda_update = self._get_beta_lamda_accumulators()
        return training_ops.apply_adam(var, m, v, 
            math_ops.cast(beta1_power, var.dtype.base_dtype), 
            math_ops.cast(beta2_power, var.dtype.base_dtype), 
            math_ops.cast(lambda_update, var.dtype.base_dtype), 
            math_ops.cast(self._beta1_t, var.dtype.base_dtype), 
            math_ops.cast(self._beta2_t, var.dtype.base_dtype), 
            math_ops.cast(self._epsilon_t, var.dtype.base_dtype), 
            grad, 
            use_locking=self._use_locking).op
    def _resource_apply_dense(self, grad, var):
        m = self.get_slot(var, "m")
        v = self.get_slot(var, "v")
        beta1_power, beta2_power, lambda_update = self._get_beta_lamda_accumulators()
        return training_ops.resource_apply_adam(var.handle, 
            m.handle, 
            v.handle, 
            math_ops.cast(beta1_power, grad.dtype.base_dtype),
            math_ops.cast(beta2_power, grad.dtype.base_dtype), 
            math_ops.cast(lambda_update, var.dtype.base_dtype), 
            math_ops.cast(self._beta1_t, grad.dtype.base_dtype), 
            math_ops.cast(self._beta2_t, grad.dtype.base_dtype), 
            math_ops.cast(self._epsilon_t, grad.dtype.base_dtype), 
            grad, use_locking=self._use_locking)

    def _get_beta_lamda_accumulators(self):
        with ops.init_scope():
            if context.executing_eagerly():
                graph = None
            else:
                graph = ops.get_default_graph()
        return (self._get_non_slot_variable("beta1_power", graph=graph), 
            self._get_non_slot_variable("beta2_power", graph=graph), 
            self._get_non_slot_variable("lambda_update", graph=graph))

    def _finish(self, update_ops, name_scope):
        #Trasnformer: constant_d_model_tensor**(-0.5)*min(global_step**(-0.5), global_step*(num_warmup_steps**(-1.5)))
        """Updates beta_power variables every n batches and incrs counter."""
        
        iter_ = self._get_iter_variable()
        constant_iter_plus = K.cast(iter_ + 1, 'float32') #constant_iter_plus = iter_+ plus 1
        print("iter", iter)
        beta1_power, beta2_power, lambda_update = self._get_beta_lamda_accumulators()
        with tf.control_dependencies(update_ops):
            with tf.colocate_with(iter_):
                def update_beta_op():
                    update_beta1 = beta1_power.assign(beta1_power * self._beta1_t, use_locking=self._use_locking)
                    update_beta2 = beta2_power.assign(beta2_power * self._beta2_t, use_locking=self._use_locking)
                    update_lambda_update = lambda_update.assign(tf.math.pow(self.constant_d_model_tensor, self.constant_minus05_tensor) * tf.math.minimum(tf.math.pow(constant_iter_plus, self.constant_minus05_tensor), constant_iter_plus*tf.math.pow(self.constant_warmup_tensor, self.constant_minus15_tensor)), use_locking=self._use_locking)
                    return tf.group(update_beta1, update_beta2, update_lambda_update)
                maybe_update_beta = tf.cond(tf.equal(tf.math.mod(iter_ + 1, self._n_t), 0), update_beta_op, tf.no_op)
                with tf.control_dependencies([maybe_update_beta]):
                    update_iter = iter_.assign(iter_ + 1, use_locking=self._use_locking)
        return tf.group(*update_ops + [update_iter, maybe_update_beta], name=name_scope)

```

**Comment by fstahlberg** [comment]
have you looked at learning rate decays in T2T (noam etc.)? The multistep optimizer should be compatible to these schemes. Btw if self.model.optimizer._lr_t results in an AttributeError are you sure that you are using the right class?

**Comment by hoangcuong2011** [comment]
@fstahlberg : Thx for your reply. 
I don't think with learning rate decays in T2T (noam etc.), I can do warm-up - a trick that is also important to train Transformer. Could you help me confirm this?

BTW I am pretty sure self.model.optimizer._lr_t results in an AttributeError (see the screenshot). My callback class inherits from tf.keras.callbacks.Callback (`class My_Callback(tf.keras.callbacks.Callback)`) and MultistepAdamOptimizer, of course inherits from tf.compat.v1.train.Adam (`class MultistepAdamOptimizer(tf.compat.v1.train.AdamOptimizer)`)
Do I miss anything here, in your opinion?
Thank you!

![image](https://user-images.githubusercontent.com/8759715/70820652-c235ca00-1e0b-11ea-992f-480c80bd02e3.png)

**Comment by fstahlberg** [comment]
You can specify the number of warm up steps in t2t.

I cannot test code at the moment, but it seems that the optimizer you are referencing is actually a `TFOptimizer`, which seems to be a Keras wrapper for native TensorFlow optimizers.

https://github.com/tensorflow/tensorflow/blob/ad6be7820bd81872d7b398df4843f6cdfba52192/tensorflow/python/keras/optimizers.py#L711

So maybe try `self.model.optimizer.optimizer._lr_t`. It is sometimes helpful to use `dir(self.model.optimizer)` in pdb to verify what you are dealing with.

**Comment by hoangcuong2011** [comment]
`self.model.optimizer.optimizer._lr_t` -> it works!!!! Thank you so much for this. This means my code is useless because changing the learning rate via Callback is much simpler to do. Hower, this is a very nice experience to learn.

Thank you so much @fstahlberg! This is a second time I owned your help (The first time is when I asked you via email about your paper On NMT Search Errors and Model Errors: Cat Got Your Tongue. Thank you!)

**Comment by hoangcuong2011** [comment]
Actually, after careful investigation I noticed that by simply changing `self.model.optimizer.optimizer._lr_t` within Callback does not work. This is because in multistep_optimizer, the graph was created and fix so that changing _lr_t wont necessarily update the value of tensor in the graph. So the only way to do that is to modify `_lr_t` inside `MultistepAdamOptimizer` class. I managed to finish this task now., btw. Thank you all, again!
